From f5ce5684ec88d186481dfcb7c5f93085f3fa8978 Mon Sep 17 00:00:00 2001
From: Luca Stefani <luca.stefani.ge1@gmail.com>
Date: Sat, 5 May 2018 12:51:42 +0200
Subject: [PATCH] Allow providing flex and bison binaries

* The prebuilt flex binary crashes due changes in newer glibc versions,
  and the prebuilt bison binary doesn't work on newer Darwin versions.
  -> Add a flag to allow the user to provide its own flex and bison binary,
     that should match the version provided by AOSP.

Usage: Add `export FLEX_EXEC=$path_to_flex` in your .bashrc/.zshrc
Example: `export FLEX_EXEC=$HOME/android/flex-2.5.39`

Change-Id: Ia93572e54accee1bb07f47f77ec8e159695b0e25
---
 cc/config/global.go | 14 ++++++++++++++
 cc/gen.go           |  4 ++--
 2 files changed, 16 insertions(+), 2 deletions(-)

diff --git a/cc/config/global.go b/cc/config/global.go
index 56de351..c393c93 100644
--- a/cc/config/global.go
+++ b/cc/config/global.go
@@ -170,6 +170,20 @@ func init() {
 		}
 		return "", nil
 	})
+
+	pctx.VariableFunc("FlexExec", func(config interface{}) (string, error) {
+		if override := config.(android.Config).Getenv("FLEX_EXEC"); override != "" {
+			return override, nil
+		}
+		return "prebuilts/misc/${HostPrebuiltTag}/flex/flex-2.5.39", nil
+	})
+
+	pctx.VariableFunc("BisonExec", func(config interface{}) (string, error) {
+		if override := config.(android.Config).Getenv("BISON_EXEC"); override != "" {
+			return override, nil
+		}
+		return "prebuilts/misc/${HostPrebuiltTag}/bison/bison", nil
+	})
 }
 
 var HostPrebuiltTag = pctx.VariableConfigMethod("HostPrebuiltTag", android.Config.PrebuiltOS)
diff --git a/cc/gen.go b/cc/gen.go
index 7a22abd..8ebd271 100644
--- a/cc/gen.go
+++ b/cc/gen.go
@@ -25,8 +25,8 @@ import (
 )
 
 func init() {
-	pctx.SourcePathVariable("lexCmd", "prebuilts/misc/${config.HostPrebuiltTag}/flex/flex-2.5.39")
-	pctx.SourcePathVariable("yaccCmd", "prebuilts/misc/${config.HostPrebuiltTag}/bison/bison")
+	pctx.SourcePathVariable("lexCmd", "${config.FlexExec}")
+	pctx.SourcePathVariable("yaccCmd", "${config.BisonExec}")
 	pctx.SourcePathVariable("yaccDataDir", "external/bison/data")
 
 	pctx.HostBinToolVariable("aidlCmd", "aidl-cpp")
-- 
2.17.0

